docs/adr/0006-json-schemas-for-messaging.md

```diff
@@ -0,0 +1,92 @@
+- [Semantic Versioning][schemaver] to effectively communicate breaking vs. non-breaking message format changes
+
+## Considered Options
+
+1. Keep existing approach
+2. Use protobuf to format/validate/document message formats
+3. Use JSON-Schemas with SemVer for doc/validation of existing JSON message formats
+
+## Decision Outcome
+
+PLACEHOLDER
+
+### [option 1] Keep existing approach
+
+Currently our documentation drifts if a message format is tweaked and the docs are not updated. Ideally
+our code review process would catch this but sometimes code subleties are not caught that alter the
+message format. Validation changes can also fail to appear in the documentation, even though what is
+considered a valid message key/value should be documented.
+
+We have to occasionally duplicate `joi` schemas between repos, which should likely be moved to `fxa-shared`
+to benefit re-use, but this does not help non-Node projects consume our messages.
+
+- Pros
+  - No additional work is needed that we don't already do.
+- Cons
+  - Our documentation occasionally drifts.
+  - It's additional work to maintain accurate `joi` schemas spread throughout the code-base _or_
+  - It's additional work to consolidate our `joi` schemas in `fxa-shared`.
+
+### [option 2] Use protobuf
+
+Using protobuf would allow us to validate and document valid message formats that would also consume
+less space and be more performant to serialize/deserialize. Validation is built-in as invalid messages
+will not deserialize, and documenation can be generated based on the protobuf spec for a message.
+
+Protobuf tooling is widely available for most langauges.
+
+- Pros
+  - High throughput messaging could benefit from higher efficiency of serializing/deserializing of messages
+  - Documentation stays in sync with message spec
+  - Validation is built-in to the message serialization/deserialization process
+- Cons
+  - Messages are in binary and not easily introspectable
+  - Existing message consumers will need to be updated to handle protobuf
+  - Migration procedure will need to be created/implemented to shift from existin JSON format
+
+### [option 3] Use JSON-Schemas
+
+JSON-Schemas can define existing FxA JSON messages used in our queue system. There are libraries in
+most languages to consume these schemas for validation purposes, and libraries exist to generate
+documenation from a JSON-Schema.
+
+[SchemaVer] can be introduced to existing FxA JSON messages as consumers ignore unknown keys, while
+consumers using the JSON-Schema can verify they can handle new message formats or if they're incompatible.
+Some of this logic is manual, as [SchemaVer] is not an official part of the JSON-Schema specification.
+
+- Pros
+  - No changes needed for existing consumers
+  - Validation can be done based on the JSON-Schema without custom code
+  - Documentation can be generated to stay in sync with message spec
```

Comment:
Do you have an example of documentation generated from a JSON-Schema? I don't find the raw schema definitions to be particularly for casual readers, which makes me wonder if we'd end up maintaining parallel "narrative" docs in addition to the schema.

Reply:
Yes, [Adobe built one](https://github.com/adobe/jsonschema2md) that produces[ output like this](https://github.com/adobe/jsonschema2md/blob/master/examples/docs/example.schema.md).

Reply:
This was one of my concerns as well. I wonder if there is a way to import the schema doc into a more narrative doc.

I haven't played with JSONSchema enough to know, is it possible to generate full API docs like we have at https://github.com/mozilla/fxa/blob/master/packages/fxa-auth-server/docs/api.md?

Reply:
The docs there are API docs, which is more like what Swagger would generate. For message formats we'd want to generate something closer to:
https://github.com/mozilla/fxa/blob/master/packages/fxa-auth-server/docs/service_notifications.md

Using the above output from the Adobe tool, we could generate a page for each message schema, then still have a narrative doc page that links to each generated message page.

Reply:
I've worked a bit with JSON Schema when working on WebExtension APIs. There are some really nice things about it. You can do things like declare a type and refer back to it using `$id`, and collect shared types or examples in a single spot, then point to them across schema files using `$ref`. Here are some relevant docs on managing large / complex schemas: https://json-schema.org/understanding-json-schema/structuring.html

> Do you have an example of documentation generated from a JSON-Schema?

Here is the schema I wrote for the Telemetry WebExtension API: https://searchfox.org/mozilla-central/source/toolkit/components/extensions/schemas/telemetry.json

And here are the docs auto-generated from the schema: https://firefox-source-docs.mozilla.org/toolkit/components/telemetry/telemetry/collection/webextension-api.html

>  I haven't played with JSONSchema enough to know, is it possible to generate full API docs like we have at https://github.com/mozilla/fxa/blob/master/packages/fxa-auth-server/docs/api.md?

There's a JSON Hyper-Schema standard as well, which sounds like what you're after. Check out this blog post from Cloudflare, which goes into detail on how they auto-generate their API docs from Hyper-Schema: https://blog.cloudflare.com/cloudflares-json-powered-documentation-generator/

Also see the list of tools for Schema and Hyper-Schema: http://json-schema.org/implementations.html